### ISSUE TITLE
Map results in degenerate signals

### ISSUE THREAD

**Issue by woozey** [hyperspy/hyperspy#1517]
I'm not sure if it is a bug or just wrong usage of `map` function, for example starting with `signal2d` and trying to use any function outputting a single value:

> In [90]:
tt = hs.signals.Signal2D(np.arange(9).reshape((3,3)))
In [91]:
ttt = tt.map(np.sum, inplace=False)
In [92]:
ttt
Out[92]:
<Signal2D, title: , dimensions: (|)>

results in a "degenerate" (  (|)  ) `signal2d` which is already not the best behaviour... Than converting it to a `base_signal` will be still a degenerate signal which creates troubles with saving/loading:

> In [93]:
ttt.set_signal_type('base_signal')
In [94]:
ttt
Out[94]:
<BaseSignal, title: , dimensions: (|)>
In [96]:
ttt.save('./datasets/test.hdf5')
In [97]:
ttt_l = hs.load('./datasets/test.hdf5')

> KeyError                                  Traceback (most recent call last)
/Users/vadim/Work_Recent/Prog/_Py/hyperspy/hyperspy/io_plugins/hdf5.py in hdfgroup2signaldict(group, lazy)
    226             axes = [i for k, i in sorted(iter(hdfgroup2dict(
--> 227                 group['_list_' + str(len(exp['data'].shape)) + '_axes'],
    228                 lazy=lazy).items()))]

> h5py/_objects.pyx in h5py._objects.with_phil.wrapper (/Users/ilan/minonda/conda-bld/work/h5py/_objects.c:2579)()

> h5py/_objects.pyx in h5py._objects.with_phil.wrapper (/Users/ilan/minonda/conda-bld/work/h5py/_objects.c:2538)()

>/Users/vadim/anaconda/lib/python3.5/site-packages/h5py/_hl/group.py in __getitem__(self, name)
    163         else:
--> 164             oid = h5o.open(self.id, self._e(name), lapl=self._lapl)
    165 

> h5py/_objects.pyx in h5py._objects.with_phil.wrapper (/Users/ilan/minonda/conda-bld/work/h5py/_objects.c:2579)()

> h5py/_objects.pyx in h5py._objects.with_phil.wrapper (/Users/ilan/minonda/conda-bld/work/h5py/_objects.c:2538)()

> h5py/h5o.pyx in h5py.h5o.open (/Users/ilan/minonda/conda-bld/work/h5py/h5o.c:3551)()

> KeyError: "Unable to open object (Object '_list_1_axes' doesn't exist)"

> During handling of the above exception, another exception occurred:

> OSError                                   Traceback (most recent call last)
<ipython-input-97-1041497ad4ee> in <module>()
----> 1 ttt_l = hs.load('./datasets/test.hdf5')

> /Users/vadim/Work_Recent/Prog/_Py/hyperspy/hyperspy/io.py in load(filenames, signal_type, stack, stack_axis, new_axis_name, lazy, **kwds)
    229             objects = [load_single_file(filename, lazy=lazy,
    230                                         **kwds)
--> 231                        for filename in filenames]
    232 
    233         if preferences.Plot.plot_on_load:

> /Users/vadim/Work_Recent/Prog/_Py/hyperspy/hyperspy/io.py in <listcomp>(.0)
    229             objects = [load_single_file(filename, lazy=lazy,
    230                                         **kwds)
--> 231                        for filename in filenames]
    232 
    233         if preferences.Plot.plot_on_load:

> /Users/vadim/Work_Recent/Prog/_Py/hyperspy/hyperspy/io.py in load_single_file(filename, signal_type, **kwds)
    275                                 reader=reader,
    276                                 signal_type=signal_type,
--> 277                                 **kwds)
    278 
    279 

> /Users/vadim/Work_Recent/Prog/_Py/hyperspy/hyperspy/io.py in load_with_reader(filename, reader, signal_type, **kwds)
    284     lazy = kwds.get('lazy', False)
    285     file_data_list = reader.file_reader(filename,
--> 286                                         **kwds)
    287     objects = []
    288 

> /Users/vadim/Work_Recent/Prog/_Py/hyperspy/hyperspy/io_plugins/hdf5.py in file_reader(filename, backing_store, lazy, **kwds)
    161         for experiment in experiments:
    162             exg = f['Experiments'][experiment]
--> 163             exp = hdfgroup2signaldict(exg, lazy)
    164             # assign correct models, if found:
    165             _tmp = {}

> /Users/vadim/Work_Recent/Prog/_Py/hyperspy/hyperspy/io_plugins/hdf5.py in hdfgroup2signaldict(group, lazy)
    228                 lazy=lazy).items()))]
    229         except KeyError:
--> 230             raise IOError(not_valid_format)
    231     exp['axes'] = axes
    232     if 'learning_results' in group.keys():

> OSError: The file is not a valid HyperSpy hdf5 file

**Comment by to266** [comment]
It's on the border of a bug and a bad use of `map`, as it's meant to iterate over the navigation space (and obviously works for more pixels):
```python
In [23]: s = hs.signals.Signal2D(np.arange(9).reshape((3,3)))

In [24]: ss = hs.stack([s,s])

In [25]: t = ss.map(np.sum, inplace=False, ragged=False)

In [26]: t
Out[26]: <Signal2D, title: Stack of , dimensions: (2|)>
```

Generally at some point we decided that a HyperSpy signal without any axes does not make sense, and hence always should include at least one ( e.g. scalar) axis. That's probably why the saving/loading cycle fails with the no-axes result. 

The `map` checked that the result scalar does not have `shape` attribute, and hence defaulted to remove all signal axes. The implicit assumption is then to only leave the navigation axes, but there were none, and hence the broken signal.

@francisco-dlp should we check in `map` that there are at least two navigation pixels? I can see that currently we don't.

**Comment by woozey** [comment]
It seem to that both together gives a problem: i.e. when nav_dim=0 for input/output and sig_dim=1 for output signals, for example, the following works nicely (referring to my example above):
> In [117]:
ddd = tt.map(np.sum, axis=0, inplace=False)
In [118]:
ddd
Out[118]:
<Signal2D, title: , dimensions: (|3)>,

and also this combination looks ok:
> In [119]:
ttts = hs.signals.BaseSignal(ttt.data)
In [120]:
ttts
Out[120]:
<BaseSignal, title: , dimensions: (|1)>

Is there a chance to check if 'map' results in a degenerate signal, than to assign the signal axis dimensionality to 1?

**Comment by francisco-dlp** [comment]
Fixed in #1518
